### ISSUE TITLE
embind: no decimals in float or double constants

### ISSUE THREAD

**Issue by Guardado** [kripken/emscripten#5565]
I may be doing something wrong but I don't get float or double const numbers to output the decimal part of the number in javascript. Using both WASM or asm.js

Here is an example:

test.cpp
```C++
#include <stdio.h>
#include <emscripten/bind.h>

using namespace emscripten;

#define NUM 4
#define PI 3.1416
#define EULER 2.7182818
const int num = 4;
const float pi = 3.1416;
const double euler = 2.7182818;

EMSCRIPTEN_BINDINGS(myconstants) {
	constant("NUM", NUM);
	constant("PI", PI);
	constant("EULER", EULER);
	constant("num", num);
	constant("pi", pi);
	constant("euler", euler);
}
```
```
$ em++ --bind test.cpp -s WASM=1 -o test.js
```

const_test.js
```javascript
var constant = require("./test.js");

constant.onRuntimeInitialized = function() {
    console.log('Init');
    console.log("NUM = " + constant.NUM);
    console.log("PI = " + constant.PI);
    console.log("EULER = " + constant.EULER);
    console.log("num = " + constant.num);
    console.log("pi = " + constant.pi);
    console.log("euler = " + constant.euler);
}
```
```
$ node const_test.js
```

Output
```
Init
NUM = 4
PI = 3
EULER = 2
num = 4
pi = 3
euler = 2
```

PI and pi should output = 3.1416
EULER and euler should output = 2.7182818

Edit: Forgot to add EULER and euler for printing.

**Comment by AlexPerrot** [comment]
Hi,
After some investigation, it seems the culprit is the call to `asGenericValue` [here](https://github.com/kripken/emscripten/blob/6dc4ac5f9e4d8484e273e4dcc554f809738cedd6/system/include/emscripten/bind.h#L1597). I suspect this should not be called for native types.

**Comment by jbencook** [comment]
I'm running into the same issue. Has anyone found a reasonable workaround for this?

**Comment by froydnj** [comment]
I will look at this tomorrow or Friday.

**Comment by froydnj** [comment]
OK, so this stems from the interplay of a couple of things:

1. We declare that `float` and `double` have a `WireType` of themselves [in `wire.h`](https://github.com/kripken/emscripten/blob/6dc4ac5f9e4d8484e273e4dcc554f809738cedd6/system/include/emscripten/wire.h#L219-L241).
2. `asGenericValue` assumes that we can convert from every `WireType` to `uintptr_t`, which doesn't really work in general (e.g. `double`, `uint64_t`).
3. `asGenericValue` further assumes that for all non-pointer types, we can accomplish this conversion to `uintptr_t` via `static_cast`, which results in truncation of floating-point values.

You can "fix" `asGenericValue` for `float` to do bitwise-reinterpretation to `uintptr_t`, leaving the `double` case aside for a moment.  But `_embind_register_constant` assumes that the bits passed along as the constant value [can be handed directly to the internal `fromWireType`](https://github.com/kripken/emscripten/blob/6dc4ac5f9e4d8484e273e4dcc554f809738cedd6/src/embind/embind.js#L2329-L2336) (distinct from the C++ templates!) for the registered type.  And the `fromWireType` for floating-point types [assumes that the wire type is the floating-point type](https://github.com/kripken/emscripten/blob/6dc4ac5f9e4d8484e273e4dcc554f809738cedd6/src/embind/embind.js#L566-L582).  Which means that you get back the floating-point constant as its bits, rather than the float you were expecting.

The above can be addressed, I think, by changing the C++ and JS side `fromWireType` and `toWireType` for `float` to do bitwise reinterpretation.  But that makes the JS side somewhat slower, and it still doesn't fix the `double` case.

You could fix both `float` and `double` by changing the C++ on-wire types to be heap-allocated structures containing a `float` or `double`, respectively.  `asGenericValue` wouldn't need to be touched.  The JS side would need to be changed more dramatically, and you'd also be incurring allocations to pass simple floating-point values around, which I don't think anybody expects.

I think the simplest solution is to create an `_embind_register_constant_float` API, so we can pass in the desired constant with no loss of information.

**Comment by Becavalier** [comment]
Already fixed by #6696.
